### PR TITLE
Fix for #71 - Hidden new properties

### PR DESCRIPTION
**Pull request by AlexeyRaga**
Addresses #71  as per issue description

### CODE REVIEW

**Comment by AlexeyRaga** [comment]
Yes. The "problem" here is (and it is another "feature" of HH-autogen) that you have constructors with parameters.

The code currently works this way (non't really know why): if the type has a constructor with parameters, then it will have `Shape.Type = Shape.Poco` and it will _only_ generate and pass constructor parameters. And it will _never_ attempt to set properties. Which is, perhaps, somehow fair because we don't know what has been already done and set by the constructor.

But if the type doesn't have a constructor, or only has a parameterless constructor,  then it has `Shape.CliMutable` and it starts setting properties.

Long story short: drop parameters in your constructors:

```fsharp
type Anaimal() =
  let mutable value = null
  member _.Value
    with get() : obj = value
    and set(v: obj) = value <- v

type Dog() =
  inherit Anaimal()
  member _.Value
    with get() : string = base.Value :?> string
    and set(v: string) = base.Value <- v

[<Fact>]
let ``test123`` () =
  Property.check <| property {
    let actual =
      GenX.auto<Dog>
      |> Gen.sample 0 1
      |> Seq.head
    actual.Value // Does not throw an exception
    |> ignore
  }
```

and this test fails.

**Comment by TysonMN** [comment]
Excellent work

**Comment by TysonMN** [comment]
I just force pushed to your branch.  I implemented a fix that only sets each property once.

What do you think?

**Comment by TysonMN** [comment]
I just force pushed again to add another test.  It is the test that tries to auto generate the type `Poodle`.  This correctness test fails for your original fix.  Ordering by increasing "type depth" makes this test pass.  Then only keeping the property with the largest "type depth" fixes the optimization test (about only setting each property once).

**Comment by TysonMN** [comment]
Ok, last force push.  I split my commits so that correctness is fixed first and then the optimization happens second.

**Comment by TysonMN** [comment]
The build says there are 99 tests (and all passed), but my Visual Studio says there are 103 tests (and all pass).  I want to fix this before merging.

File: src/Hedgehog.Experimental/Gen.fs [hunk 1]
```diff
@@ -597,8 +597,16 @@ module GenX =
...
+              let rec loop (t: Type) depth =
+                if t = null
+                then depth
+                else loop t.BaseType (depth + 1)
+              loop sm.MemberInfo.DeclaringType 0
             shape.Properties
-            |> Seq.toList
+            |> Array.toList
+            |> List.groupBy (fun p -> p.MemberInfo.Name)
+            |> List.map (fun (_, ps) -> ps |> List.sortByDescending getDepth |> List.head)
```
Comment (AlexeyRaga) [inline src/Hedgehog.Experimental/Gen.fs:609]: I wonder if `snd >> List.sortByDescending getDepth >> head` can be replaced with `snd >> Seq.maxBy getDepth` for clarity.

Also for efficiency we can skip calling to `getDepth` if there is only one member in the group.

Sorry if it looks like nitpicking, but I see big projects with thousands of tests, and unnecessary work does add up to a significant time in running tests / CI...

File: src/Hedgehog.Experimental/Hedgehog.Experimental.fsproj [hunk 2]
```diff
@@ -35,7 +35,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Update="FSharp.Core" Version="4.3.4" />
+    <PackageReference Update="FSharp.Core" Version="4.6.2" />
```
[the remark this answers is not in the document]
Comment (AlexeyRaga) [reply src/Hedgehog.Experimental/Hedgehog.Experimental.fsproj:38]: 1. Because the test project already uses `4.6.2` so at least be consistent within the solution
2. Because `Seq.tryExactlyOne` was introduced years ago but still after even more ancient `4.3.4`

[the remark this answers is not in the document]
Comment (TysonMN) [reply src/Hedgehog.Experimental/Hedgehog.Experimental.fsproj]: These are not good enough reasons to change this.  It is fine for the test code to use the latest version of `FSharp.Core`.  That allows us to use new things like `Seq.tryExactlyOnce`.  Increasing the version of `FSharp.Core` places stricter requirements on users.

It is easy to implement your optimization while continuing to use version `4.3.4` of `FSharp.Core`.  I just forced pushed to the branch after switching to this code:

```fsharp
 |> List.map (snd >> function
                     | [p] -> p
                     | ps -> ps |> List.sortByDescending getDepth |> List.head)
```

[the remark this answers is not in the document]
Comment (AlexeyRaga) [reply src/Hedgehog.Experimental/Hedgehog.Experimental.fsproj]: It is not latest BTW, it is at least 3 years old. Latest is `6.x.x`.
